modules/lfstransfer/backend/backend.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package backend
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"code.gitea.io/gitea/modules/json"
+	"code.gitea.io/gitea/modules/lfs"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/charmbracelet/git-lfs-transfer/transfer"
+)
+
+// Version is the git-lfs-transfer protocol version number.
+const Version = "1"
+
+// Capabilities is a list of Git LFS capabilities supported by this package.
+var Capabilities = []string{
+	"version=" + Version,
+	"locking",
+}
+
+var _ transfer.Backend = &GiteaBackend{}
+
+// GiteaBackend is an adapter between git-lfs-transfer library and Gitea's internal LFS API
+type GiteaBackend struct {
+	ctx    context.Context
+	server *url.URL
+	op     string
+	token  string
+	logger transfer.Logger
+}
+
+func New(ctx context.Context, repo, op, token string, logger transfer.Logger) (transfer.Backend, error) {
+	// runServ guarantees repo will be in form [owner]/[name].git
+	server, err := url.Parse(setting.LocalURL)
+	if err != nil {
+		return nil, err
+	}
+	server = server.JoinPath(repo, "info/lfs")
```

Comment:
Should we request web route or another internal route?

Reply:
I don't see any issues with using the public routes, instead of making exact copies for internal use. In fact we save a lot of code duplication because we can just defer the access control logic to the public API code: to it, this looks just like it was a request coming directly from the user because of the `token`.

Reply:
I don't think this can work. `setting.LocalURL` is for private routes which have different routes from `setting.AppURL`. Commonly, `setting.LocalURL` should be `localhost`. Maybe you need to use `setting.AppURL`.

Reply:
@lunny I mean, it does work. It's probably because the public routes don't discriminate whether the request comes from `localhost` or from an external ip.

Do you think it should be changed to `AppURL`? I think that will work just fine as well, but one downside would be that all of this will probably stop working if the user doesn't define an `AppURL` i.e. to use Gitea as purely a git server over SSH, without a web interface. Of course, that depends on whether Gitea aims to support such a setup, but situations like this is pretty much what the LFS SSH protocol was created for.

Reply:
(To be clear, it will still work if the user defines an `AppURL` but just keeps it private to the network, like the creator of the issue wanted to. But it won't if they don't define it at all.)

Reply:
`LocalURL` means internal routes. ref https://github.com/go-gitea/gitea/blob/main/routers/private/internal.go#L51 I couldn't find the router for `<repo>/info/lfs` there.
Yes, `AppURL` means public routers. That's why I have asked why not create a new router in https://github.com/go-gitea/gitea/blob/main/routers/private/internal.go#L51 I

Reply:
@lunny the LFS API router is here:
https://github.com/go-gitea/gitea/blob/e0c27e59968f0ecc5e433617f303520fb765b06a/routers/web/web.go#L1584

As you can see, it doesn't call the `CheckInternalToken` middleware, so it can be accessed both publically and privately (authentication happens later using the user's JWT token).

> That's why I have asked why not create a new router in https://github.com/go-gitea/gitea/blob/main/routers/private/internal.go#L51 I

My apologies, but I don't quite get what you mean. If `AppURL` seems more appropriate we can change the calls to use that, and everything will still work. Creating a new router for the same API would just mean there are two 
listeners registered for the same paths. I'm not sure what the behaviour would be, maybe it would select one of them randomly or maybe the one registered later would always be called. But to me that seems useless at  best, and buggy at worst.